rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+# Feature Name: `stageless`
+
+## Summary
+
+The existing stage boundaries result in a large number of complications: preventing us from operating across stages.
+Moreover, several related features (run criteria, states and fixed timesteps) are rather fragile and overly complex.
+This tangled mess needs to be holistically refactored: simplifying the scheduling model down into a single `Schedule`, moving towards an intent-based configuration style, taking full advantage of labels and leveraging exclusive systems to handle complex logic.
+
+## Motivation
+
+The fundamental challenges with the current stage-driven scheduling abstraction are [numerous](https://github.com/bevyengine/bevy/discussions/2801).
+Of particular note:
+
+- Dependencies do not work across stages.
+- States do not work across stages.
+- Plugins can add new, standalone stages.
+- Run criteria (including states and fixed timestep run criteria!) cannot be composed.
+- Run criteria cannot be reused across states or stages.
+- The stack-driven state model is overly elaborate and does not enable enough use cases to warrant its complexity.
+- The architecture for turn-based games and other even slightly unusual patterns is not obvious.
+- We cannot easily implement a builder-style strategy for configuring systems.
+
+Unfortunately, all of these problems are deeply interwoven. Despite our best efforts, fixing them incrementally at either the design or implementation stage is impossible, as it results in myopic architecture choices and terrible technical debt.
+
+## User-facing explanation

```

Comment:
From discussion with @plof27, we have three tools for disabling systems, each with their own use cases:

> 1. Schedule alteration: Systems that may or may not be run in the game (ever), and runtime pipelining optimizations 
> 2. Exclusive systems: More complex, looping schedules without nasty overhead (but no parallel execution without using unsafe)
> 3. Run criteria: Nice, automatic parallel scheduling without any custom implementation required

We need to explain the differences and use cases up-front to users.

Reply:
I think we should use different language for "disabling" in each context. RC handle conditional execution per schedule iteration, schedule mod handles "permanent" insert/remove, and exclusive systems are sledgehammers for doing arbitrary things.

> 1. Schedule alteration: Systems that may or may not be run in the game (ever), and runtime pipelining optimizations.

Not sure what "runtime pipelining optimization" means, but I believe this scenario refers to _removing_ systems, which translates to a graph rebuild. Otherwise, it'd pretty much be no different than tacking on an RC that always returns `false`.

If we have systems that could never run based on the registered components, our options are:
- Do nothing.
- Return a schedule build error.
- Auto-remove them from any graphs when built. 

That last one would be us marking systems that ask for unregistered types, during (re)init. We have to first calculate the graph's transitive closure to remove spurious edges anyway, so we'd just remove the nodes of the marked systems then too.

> 2. Exclusive systems [that run schedules inside]

Users being able to put a condition for running a nested schedule inside the enclosing system is kinda just coincidence.  Exclusive systems are like sledgehammers. The real pattern here is encapsulating non-DAG stuff inside a system:
- Does user want to run a group of systems in a loop (in the same schedule iteration)?
  - (e.g. fixed timestep, processing an event queue)
- Does user want to run a group of systems ad-hoc (at the next available opportunity)?
  - (e.g. FSM transitions)

Also, technically, this pattern is only limited to exclusive systems rn because `Schedule::run(world: &mut World)` is the only reasonable fn signature.

If we look at a sub-graph, its access is just the union of those inside it. So if some sub-graph was atomic and contained some system fetching `&mut World` (such as a flush), it would naturally become "exclusive" itself and block other stuff from running. Conversely, not every sub-graph needs to borrow the whole world.

The problem is that encapsulating systems don't have a good way to report what their enclosed sub-graphs are borrowing. We just know `&mut World` always works.

I focused on "sub-graph" because all `Schedule` is, really, is a tuple of sub-graph + system storage + executor.

> Run criteria: Nice, automatic parallel scheduling without any custom implementation required.

RC are largely irrelevant to the graph construction. The builder only looks at them when it checks the graph for ambiguities and spurious edges.

RC are pretty much if-statements that don't get in the way of change detection, so any "if <read-only check>, run system" is a valid use case.

Reply:
Schedule commands have been moved to Future Work. IMO the distinction between running custom schedules and run criteria is very apparent and does not warrant further elaboration.